lib/TPP/Dialect/Tpp/TppUtils.cpp

```diff
@@ -334,24 +334,67 @@ static bool isBinaryOp(linalg::GenericOp linalgOp) {
   return false;
 }
 
-// Return true if the operands are all the same.
-static bool allOperandsHaveSameType(linalg::GenericOp linalgOp) {
-  Type operandType = linalgOp->getOperand(0).getType();
-  for (Value operand : linalgOp->getOperands()) {
-    Type currentType = operand.getType();
-    if (currentType != operandType)
+// Returns true if all the operands of the current linalgOp are shape type and
+// have the same statically-known shape. For example: A: memref<56x32xf32,
+// strided<[32, 1], offset: ?>> B: memref<56x32xf32> allOperandsHaveSameShape(A,
+// B) return true. C: memref<1x32xf32> allOperandsHaveSameShape(A, C) return
+// false.
+bool allOperandsHaveSameShapeAndStrides(TypeRange types) {
+  assert(types.size() > 1 && "expect one or more types");
+  if (!types[0].isa<MemRefType>())
+    return false;
+  auto firstOperandType = types[0].cast<MemRefType>();
+
+  // Step1. Validate rank.
+  int64_t rankOperand = firstOperandType.getRank();
+  for (Type currentType : types) {
+    if (!currentType.isa<MemRefType>())
+      return false;
+    if (currentType.cast<MemRefType>().getRank() != rankOperand)
+      return false;
+  }
+
+  // Step2. Validate shape, and strides.
+  // Get stride and offset for the first operand.
+  int64_t offsetFirstOperand = 0;
+  SmallVector<int64_t> stridesFirstOperand;
+  if (failed(getStridesAndOffset(firstOperandType, stridesFirstOperand,
+                                 offsetFirstOperand)))
+    return false;
+  ArrayRef<int64_t> shapeOperand = firstOperandType.getShape();
+  for (Type currentOperandType : types) {
+    // Compare the shape.
+    ArrayRef<int64_t> shapeCurrentOperand =
+        currentOperandType.cast<MemRefType>().getShape();
+    if (shapeCurrentOperand != shapeOperand)
+      return false;
+    int64_t offsetCurrentOperand = 0;
+    SmallVector<int64_t> stridesCurrentOperand;
+    // Compare the strides.
+    if (failed(getStridesAndOffset(currentOperandType.cast<MemRefType>(),
+                                   stridesCurrentOperand,
+                                   offsetCurrentOperand)))
+      return false;
+    if (stridesFirstOperand != stridesCurrentOperand)
       return false;
   }
   return true;
 }
 
+static bool allOperandsHaveSameShapeAndStrides(linalg::GenericOp linalgOp) {
+  SmallVector<Type> types;
+  for (Value operand : linalgOp->getOperands())
+    types.push_back(operand.getType());
+  return allOperandsHaveSameShapeAndStrides(types);
```

Comment:
Can't you just pass `getOperands` to the function and capture as a reference?

You probably wouldn't need the two functions, then.

Reply:
I am passing a typeRange, getOperands return a ValueRange.

Reply:
Perhaps a helper from `ValueRange` to `TypeRange` then?

Reply:
What exactly do you want to achieve here? Remove `allOperandsHaveSameShapeAndStrides(linalg::GenericOp linalgOp)`?

Reply:
For starters, yes. Any such special wrappers pollute the file with similar code that gets hard to maintain later.

This is true for the rest of the code, too. So, as you go through the code, if you find the same pattern of super-specialized functions, try to generalize them and move up to some common place, and as a last resort, TppUtils (we should also move that file to a more generic place, it's not specific to the tpp dialect any more).

Reply:
I think I've seen a helper function that extracts type ranges from value ranges in MLIR upstream...

Reply:
remember the name?

Reply:
Yes, I updated my previous comment. Turns out is just `ValueRange.getTypes()`.